File: backend/src/cron.py
"""Strict five-field cron parsing and UTC occurrence calculation."""
from datetime import datetime, timedelta, timezone

RANGES = ((0,59),(0,23),(1,31),(1,12),(0,6))

class CronError(ValueError): pass

def _field(text, low, high):
    values=set()
    for part in text.split(','):
        base, sep, step_text=part.partition('/')
        try: step=int(step_text) if sep else 1
        except ValueError: raise CronError("cron step must be an integer")
        if step < 1: raise CronError("cron step must be positive")
        if base=='*': start,end=low,high
        elif '-' in base:
            try: start,end=map(int,base.split('-',1))
            except ValueError: raise CronError("invalid cron range")
        else:
            try: start=end=int(base)
            except ValueError: raise CronError("invalid cron value")
        if start<low or end>high or start>end: raise CronError(f"cron value must be between {low} and {high}")
        values.update(range(start,end+1,step))
    return values

def parse(expression):
    parts=expression.strip().split()
    if len(parts)!=5: raise CronError("cron expression must contain exactly five fields")
    return tuple(_field(part,*bounds) for part,bounds in zip(parts,RANGES))

def matches(expression, moment):
    fields=parse(expression); moment=moment.astimezone(timezone.utc)
    return all(value in values for value,values in zip((moment.minute,moment.hour,moment.day,moment.month,(moment.weekday()+1)%7),fields))

def next_run(expression, after):
    candidate=after.astimezone(timezone.utc).replace(second=0,microsecond=0)+timedelta(minutes=1)
    for _ in range(366*24*60*5):
        if matches(expression,candidate): return candidate
        candidate += timedelta(minutes=1)
    raise CronError("no occurrence found within five years")

File: frontend/lib/cron.ts
export type CronClause = {
  kind: 'wildcard' | 'value' | 'range';
  start?: number;
  end?: number;
  step: number;
};
export type CronField = CronClause[];
export type CronSchedule = {
  minute: CronField;
  hour: CronField;
  dayOfMonth: CronField;
  month: CronField;
  dayOfWeek: CronField;
};

export const CRON_FIELDS = [
  { key: 'minute', label: 'Minute', min: 0, max: 59 },
  { key: 'hour', label: 'Hour', min: 0, max: 23 },
  { key: 'dayOfMonth', label: 'Day of month', min: 1, max: 31 },
  { key: 'month', label: 'Month', min: 1, max: 12 },
  { key: 'dayOfWeek', label: 'Weekday', min: 0, max: 6 },
] as const;

function parseField(text: string, min: number, max: number): CronField {
  if (!text) throw new Error('A cron field cannot be empty.');
  return text.split(',').map((part) => {
    const pieces = part.split('/');
    if (pieces.length > 2 || !pieces[0]) throw new Error(`Invalid schedule field: ${text}`);
    const step = pieces.length === 2 ? Number(pieces[1]) : 1;
    if (!Number.isInteger(step) || step < 1) throw new Error('Steps must be positive whole numbers.');
    const base = pieces[0];
    if (base === '*') return { kind: 'wildcard', step };
    if (base.includes('-')) {
      const values = base.split('-').map(Number);
      if (values.length !== 2 || values.some((v) => !Number.isInteger(v))) throw new Error(`Invalid range: ${base}`);
      if (values[0] < min || values[1] > max || values[0] > values[1]) throw new Error(`Values must be between ${min} and ${max}.`);
      return { kind: 'range', start: values[0], end: values[1], step };
    }
    const value = Number(base);
    if (!Number.isInteger(value) || value < min || value > max) throw new Error(`Values must be between ${min} and ${max}.`);
    return { kind: 'value', start: value, step };
  });
}

export function parseCron(expression: string): CronSchedule {
  const parts = expression.trim().split(/\s+/);
  if (parts.length !== 5) throw new Error('A schedule must contain five fields.');
  const parsed = CRON_FIELDS.map((field, index) => parseField(parts[index], field.min, field.max));
  return Object.fromEntries(CRON_FIELDS.map((field, index) => [field.key, parsed[index]])) as CronSchedule;
}

function serializeField(clauses: CronField, min: number, max: number): string {
  if (!clauses.length) throw new Error('Each field needs at least one choice.');
  return clauses.map((clause) => {
    const step = Number(clause.step);
    if (!Number.isInteger(step) || step < 1) throw new Error('Steps must be positive whole numbers.');
    const start = Number(clause.start);
    const end = Number(clause.end);
    let base: string;
    if (clause.kind === 'wildcard') base = '*';
    else if (!Number.isInteger(start) || start < min || start > max) throw new Error(`Values must be between ${min} and ${max}.`);
    else if (clause.kind === 'value') base = String(start);
    else {
      if (!Number.isInteger(end) || end < start || end > max) throw new Error(`Ranges must be between ${min} and ${max}.`);
      base = `${start}-${end}`;
    }
    return `${base}${step === 1 ? '' : `/${step}`}`;
  }).join(',');
}

export function serializeCron(schedule: CronSchedule): string {
  return CRON_FIELDS.map((field) => serializeField(schedule[field.key], field.min, field.max)).join(' ');
}

function fieldDescription(field: CronField, label: string): string {
  return field.map((clause) => {
    const base = clause.kind === 'wildcard' ? `every ${label.toLowerCase()}` : clause.kind === 'value' ? `${label.toLowerCase()} ${clause.start}` : `${label.toLowerCase()}s ${clause.start} through ${clause.end}`;
    return clause.step > 1 ? `${base}, every ${clause.step}` : base;
  }).join(' or ');
}

export function describeCron(scheduleOrExpression: CronSchedule | string): string {
  const schedule = typeof scheduleOrExpression === 'string' ? parseCron(scheduleOrExpression) : scheduleOrExpression;
  return `UTC: ${CRON_FIELDS.map((field) => fieldDescription(schedule[field.key], field.label)).join('; ')}.`;
}
